### PR TITLE
Single table for file attachments

### PR DESCRIPTION
**Pull request by SchrodingersGat**
Refactoring how "attachments" are stored, moving to a single table approach.

### Tasks

- [x] Create new table
- [x] Data migration
- [x] Remove old tables
- [x] Expose new model to API
- [x] Add a mixin class for models which will have attachments associated with them
- [x] Mixin class must provide method for accessing attachments
- [x] Expose a list of models we wish to associate attachments with
- [x] Remove old API endpoints
- [x] Update attachment tables in PUI
- [x] Update attachment tables in CUI
- [x] Update unit tests
- [x] Extract user information on upload
- [x] Unit test for data migration
- [x] Unit tests for `InvenTreeAttachmentMixin` class mixin
- [x] Work out how to determine user permissions in backend (i.e on file upload)
- [x] Work out how to determine user permissions in user interface
- [x] Update python bindings - https://github.com/inventree/inventree-python/pull/234
- [x] Update mobile app - https://github.com/inventree/inventree-app/pull/505

### Related Issues

- Closes https://github.com/inventree/InvenTree/issues/5429
- Closes https://github.com/inventree/InvenTree/issues/5703

### CODE REVIEW

File: src/backend/InvenTree/common/models.py [hunk 1]
```diff
@@ -3062,3 +3062,116 @@ def after_custom_unit_updated(sender, instance, **kwargs):
...
+    link = InvenTree.fields.InvenTreeURLField(
+        blank=True,
+        null=True,
+        verbose_name=_('Link'),
+        help_text=_('Link to external URL'),
+    )
+
+    comment = models.CharField(
+        blank=True,
+        max_length=100,
```
Comment (matmair) [inline src/backend/InvenTree/common/models.py:3129]: Can we set this to a length of 250? 100 is a bit uncommon

File: src/backend/InvenTree/common/models.py [hunk 2]
```diff
@@ -3062,3 +3062,116 @@ def after_custom_unit_updated(sender, instance, **kwargs):
...
+        verbose_name=_('Comment'),
+        help_text=_('Attachment comment'),
+    )
+
+    user = models.ForeignKey(
```
Comment (matmair) [inline src/backend/InvenTree/common/models.py:3134]: Can we make this more descriptive? Maybe upload_user?

File: src/backend/InvenTree/common/models.py [hunk 3]
```diff
@@ -3062,3 +3062,116 @@ def after_custom_unit_updated(sender, instance, **kwargs):
...
+
+    upload_date = models.DateField(
+        auto_now_add=True,
+        null=True,
+        blank=True,
+        verbose_name=_('Upload date'),
+        help_text=_('Date the file was uploaded'),
+    )
+
+    file_size = models.PositiveIntegerField(
```
Comment (matmair) [inline src/backend/InvenTree/common/models.py:3151]: Question: Is this the upload size or the current size,

**Review by matmair — APPROVED** [review]

File: src/backend/InvenTree/common/models.py (hunk 1 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/common/models.py]: Done.

File: src/backend/InvenTree/common/models.py (hunk 2 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/common/models.py]: Sure, I have renamed to "upload_user"

File: src/backend/InvenTree/common/models.py (hunk 3 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/common/models.py:3201]: This is the actual size on disk of the uploaded file. It is a "read only" field calculated when the instance is saved.

**Comment by SchrodingersGat** [comment]
Once we [squash migrations](https://github.com/inventree/InvenTree/issues/5015) this will result in a lot of code reduction. Not so at the moment...

File: src/backend/InvenTree/company/migrations/0034_manufacturerpart.py [hunk 4]
```diff
@@ -22,6 +22,7 @@ class Migration(migrations.Migration):
             ],
             options={
                 'unique_together': {('part', 'manufacturer', 'MPN')},
+                'verbose_name': 'Manufacturer Part',
```
Comment (matmair) [inline src/backend/InvenTree/company/migrations/0034_manufacturerpart.py:25]: Q: Changing migrations like this after the fact is a bit unorthodox - is there a reason?

File: src/frontend/src/defaults/formatters.tsx [hunk 5]
```diff
@@ -117,7 +117,23 @@ export function formatPriceRange(
   )}`;
 }
 
-interface RenderDateOptionsInterface {
+/*
+ * Format a file size (in bytes) into a human-readable format
+ */
+export function formatFileSize(size: number) {
+  const suffixes: string[] = ['B', 'KB', 'MB', 'GB'];
```
Comment (matmair) [inline src/frontend/src/defaults/formatters.tsx:124]: Comment (sarcastic, no change needed): Let's get ready for the gibibyte discussion

**Review by matmair — APPROVED** [review]
Great work, complicated migrations!

File: src/backend/InvenTree/company/migrations/0034_manufacturerpart.py (hunk 4 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/company/migrations/0034_manufacturerpart.py:25]: This model never had a "verbose_name" defined - which means it doesn't get exposed to the translation layer. Adding it in requires a new migration, although the "verbose_name" does not touch the database. So this is a relatively safe "hack"

File: src/backend/InvenTree/users/models.py [hunk 6]
```diff
@@ -344,6 +334,7 @@ def get_ruleset_ignore():
             'admin_logentry',
             'contenttypes_contenttype',
             # Models which currently do not require permissions
+            'common_attachment',
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply src/backend/InvenTree/users/models.py:337]: Yes, you need  "part" permissions to upload an attachment against a "part" instance for example.

File: src/frontend/src/tables/general/AttachmentTable.tsx [hunk 7]
```diff
@@ -206,20 +210,35 @@ export function AttachmentTable({
...
+    ];
+  }, []);
+
   const tableActions: ReactNode[] = useMemo(() => {
     return [
       <ActionButton
         key="add-attachment"
         tooltip={t`Add attachment`}
-        hidden={!allowEdit}
+        hidden={!user.hasAddPermission(model_type)}
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply src/frontend/src/tables/general/AttachmentTable.tsx:241]: tableActions are already memoized based on the [user] prop

File: tasks.py [hunk 8]
```diff
@@ -450,6 +450,7 @@ def migrate(c):
     manage(c, 'makemigrations')
     manage(c, 'runmigrations', pty=True)
     manage(c, 'migrate --run-syncdb')
+    manage(c, 'remove_stale_contenttypes --include-stale-apps --no-input', pty=True)
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply tasks.py:453]: It is not "needed" but it is a good place to ensure that old content types are pruned. And `migrate` is run sparingly anyway

[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply tasks.py:453]: When exposing the "permissions" to the `/api/user/roles/` endpoint I noted that there were a *lot* of out-dated content types which still were kept in the database.

File: src/backend/InvenTree/InvenTree/models.py [hunk 9]
```diff
@@ -503,200 +502,35 @@ class Meta:
...
-            ';',
-            '/',
-            '\\',
-            '|',
-            '?',
-            '*',
-            '%',
-            '~',
-            '`',
-        ]
```
Comment (matmair) [inline src/backend/InvenTree/InvenTree/models.py:661]: Q: @SchrodingersGat where was this security feature moved? I am not able to find it in the diff

Comment (SchrodingersGat) [reply src/backend/InvenTree/InvenTree/models.py:661]: These are largely covered by django's built in file name validators.

Comment (matmair) [reply src/backend/InvenTree/InvenTree/models.py:661]: Are you sure/have you verified this? We had a bunch of vulns in this specific area in the past with length overflows and target directory manipulation and now that huntr is basically dead we wont get a free security audit

Comment (SchrodingersGat) [reply src/backend/InvenTree/InvenTree/models.py:661]: I will add some more checking in to cover this

**Comment by SchrodingersGat** [comment]
Getting close now but stuck on some strange errors in the data migration testing. 

@wolflu05 @matmair any suggestions welcome here :)

**Comment by matmair** [comment]
Looks like the migrations might be tainted? Tough I can not see anything modifying the DB state outside of migrations.

The problem I see is that the error is raised in core Django itself so guarding against it might be difficult/impossible.

**Comment by SchrodingersGat** [comment]
@matmair all passing now, ready for merge if you're happy :)

File: src/backend/InvenTree/InvenTree/fields.py [hunk 10]
```diff
@@ -33,7 +33,7 @@ def __init__(self, **kwargs):
 
     def run_validation(self, data=empty):
         """Override default validation behaviour for this field type."""
-        strict_urls = get_global_setting('INVENTREE_STRICT_URLS', True, cache=False)
+        strict_urls = get_global_setting('INVENTREE_STRICT_URLS', cache=False)
```
Comment (matmair) [inline src/backend/InvenTree/InvenTree/fields.py:36]: Q: Why is this changed?

File: src/backend/InvenTree/plugin/registry.py [hunk 11]
```diff
@@ -786,7 +786,7 @@ def calculate_plugin_hash(self):
 
         for k in self.plugin_settings_keys():
             try:
-                val = get_global_setting(k, False, create=False)
+                val = get_global_setting(k)
```
Comment (matmair) [inline src/backend/InvenTree/plugin/registry.py:789]: Q: Why were the defaults removed?

**Review by matmair — APPROVED** [review]
LGTM; Great effort!

**Comment by matmair** [comment]
One additonal comment: I do not think https://github.com/inventree/InvenTree/issues/5429 is really covered by this but would be closed - wdyt?

File: src/backend/InvenTree/InvenTree/fields.py (hunk 10 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/InvenTree/fields.py:36]: The default value filters through from the definition of the setting itself, and does not need to be passed here.

File: src/backend/InvenTree/InvenTree/version.py [hunk 12]
```diff
@@ -53,13 +53,13 @@ def checkMinPythonVersion():
...
-    return get_global_setting('INVENTREE_INSTANCE', '')
+    return get_global_setting('INVENTREE_INSTANCE')
 
 
 def inventreeInstanceTitle():
     """Returns the InstanceTitle for the current database."""
-    if get_global_setting('INVENTREE_INSTANCE_TITLE', False):
-        return get_global_setting('INVENTREE_INSTANCE', 'InvenTree')
+    if get_global_setting('INVENTREE_INSTANCE_TITLE'):
+        return get_global_setting('INVENTREE_INSTANCE')
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply src/backend/InvenTree/InvenTree/version.py:62]: As above, duplicating the default value here is not required.

File: src/backend/InvenTree/report/tests.py [hunk 13]
```diff
@@ -502,7 +502,7 @@ def run_print_test(self, qs, model_type, label: bool = True):
             },
             expected_code=201,
             max_query_time=15,
-            max_query_count=1000,  # TODO: Should look into this
+            max_query_count=500 * len(qs),
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply src/backend/InvenTree/report/tests.py:505]: The label printing is pretty inefficient, but there's a *lot* of work required to profile exactly why

**Comment by SchrodingersGat** [comment]
With regard to #5429 - we can add the "tags" mixin to the new attachment model, and then the tags can be used however

**Comment by matmair** [comment]
Thank you for the addition!
